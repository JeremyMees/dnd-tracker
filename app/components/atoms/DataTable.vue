<script setup lang="ts">
import type { ColumnDef, PaginationState, Row, SortingState, TableOptions } from '@tanstack/vue-table'
import { FlexRender, getCoreRowModel, getExpandedRowModel, useVueTable } from '@tanstack/vue-table'

const emit = defineEmits<{
  remove: [number[]]
  invalidate: []
}>()

const props = defineProps<{
  columns: ColumnDef<any, any>[]
  data: any[]
  loading: boolean
  options?: Partial<TableOptions<any>>
  emptyMessage?: string
  permission?: boolean | ((item: any) => Promise<boolean>)
  expandedMarkup?: (row: Row<any>) => VNode
}>()

const globalFilter = ref<string>('')
const sorting = ref<SortingState>(props.options?.initialState?.sorting || [])
const pagination = ref<PaginationState>({ pageIndex: 0, pageSize: 10 })
const rowSelectionPermissions = ref<Record<string, boolean>>({})

// Convert 0-based to 1-based for Radix
const internalPage = computed({
  get: () => pagination.value.pageIndex + 1,
  set: (value: number) => pagination.value.pageIndex = value - 1,
})

const selectedRowLength = computed(() => table.getSelectedRowModel().rows.length)

watch(
  () => props.data?.length,
  async (length) => {
    if (length) await fetchPermissions()
  },
  { immediate: true },
)

const table = useVueTable({
  ...props.options,
  data: computed(() => props.data),
  columns: props.columns,
  manualPagination: true,
  manualSorting: true,
  manualFiltering: true,
  enableRowSelection: (row: Row<any>) => rowSelectionPermissions.value[row.original.id] ?? false,
  getCoreRowModel: getCoreRowModel(),
  getExpandedRowModel: getExpandedRowModel(),
  getRowId: row => row.id,
  onSortingChange: (updaterOrValue) => {
    valueUpdater(updaterOrValue, sorting)
    pagination.value.pageIndex = 0
    emit('invalidate')
  },
  onGlobalFilterChange: (updaterOrValue) => {
    valueUpdater(updaterOrValue, globalFilter)
    pagination.value.pageIndex = 0
    emit('invalidate')
  },
  state: {
    get globalFilter() { return globalFilter.value },
    get pagination() { return pagination.value },
    get sorting() { return sorting.value },
  },
})

defineExpose({ vueTable: table })

async function fetchPermissions() {
  const permissions: Record<string, boolean> = {}

  for (const item of props.data) {
    if (typeof props.permission === 'boolean') permissions[item.id] = props.permission
    else if (typeof props.permission === 'function') permissions[item.id] = await props.permission(item)
    else permissions[item.id] = true
  }

  rowSelectionPermissions.value = permissions
}
</script>

<template>
  <div class="flex flex-col gap-2">
    <div class="flex flex-col-reverse sm:flex-row gap-2 justify-between items-center">
      <FormKit
        :value="table.getState().globalFilter"
        type="search"
        name="search"
        prefix-icon="tabler:search"
        outer-class="$remove:mb-4 w-full sm:w-auto"
        inner-class="$remove:border-background $remove:bg-muted border-secondary bg-secondary/50"
        @input="table.setGlobalFilter($event)"
      />
      <div class="flex justify-end w-full sm:w-auto">
        <slot name="top" />
      </div>
    </div>

    <div class="rounded-lg border-4 border-secondary bg-secondary/50">
      <UiTable>
        <UiTableHeader>
          <UiTableRow
            v-for="headerGroup in table.getHeaderGroups()"
            :key="headerGroup.id"
          >
            <UiTableHead
              v-for="header in headerGroup.headers"
              :key="header.id"
              :data-pinned="header.column.getIsPinned()"
              :class="cn(
                { 'sticky bg-background/95': header.column.getIsPinned() },
                header.column.getIsPinned() === 'left' ? 'left-0' : 'right-0',
                header.column.getCanSort() ? 'cursor-pointer select-none' : '',
              )"
              @click="header.column.getToggleSortingHandler()?.($event)"
            >
              <ClientOnly>
                <div
                  class="flex items-center gap-2 w-fit"
                  :class="{
                    'bg-muted rounded-lg p-2 transition-all duration-300 text-foreground': header.column.getIsSorted(),
                  }"
                >
                  <FlexRender
                    v-if="!header.isPlaceholder"
                    :render="header.column.columnDef.header"
                    :props="header.getContext()"
                  />
                  <Icon
                    v-if="header.column.getIsSorted()"
                    :name="`tabler:sort-${header.column.getIsSorted() === 'asc' ? 'ascending' : 'descending'}`"
                    class="size-4"
                  />
                </div>

                <template #fallback>
                  <FlexRender
                    v-if="!header.isPlaceholder"
                    :render="header.column.columnDef.header"
                    :props="header.getContext()"
                  />
                </template>
              </ClientOnly>
            </UiTableHead>
          </UiTableRow>
        </UiTableHeader>

        <ClientOnly>
          <UiTableBody>
            <template v-if="table.getRowModel().rows?.length">
              <template
                v-for="row in table.getRowModel().rows"
                :key="row.id"
              >
                <UiTableRow :data-state="row.getIsSelected() && 'selected'">
                  <UiTableCell
                    v-for="cell in row.getVisibleCells()"
                    :key="cell.id"
                    :data-pinned="cell.column.getIsPinned()"
                    :class="cn(
                      { 'sticky bg-background/95': cell.column.getIsPinned() },
                      cell.column.getIsPinned() === 'left' ? 'left-0' : 'right-0',
                    )"
                  >
                    <FlexRender
                      :render="cell.column.columnDef.cell"
                      :props="cell.getContext()"
                    />
                  </UiTableCell>
                </UiTableRow>
                <UiTableRow v-if="expandedMarkup && row.getIsExpanded()">
                  <UiTableCell :colspan="row.getAllCells().length">
                    <FlexRender :render="expandedMarkup(row)" />
                  </UiTableCell>
                </UiTableRow>
              </template>
            </template>

            <slot
              v-else-if="loading"
              name="loading"
            />

            <UiTableRow v-else>
              <UiTableCell
                data-test-empty
                :colspan="columns.length"
                class="h-24 text-center text-muted-foreground"
              >
                {{ emptyMessage || '' }}
              </UiTableCell>
            </UiTableRow>
          </UiTableBody>

          <template #fallback>
            <UiTableBody>
              <slot name="loading" />
            </UiTableBody>
          </template>
        </ClientOnly>
      </UiTable>

      <div
        :class="[selectedRowLength ? 'flex-col md:flex-row justify-between' : 'justify-end']"
        class="p-4 border-t border-secondary flex items-center gap-2"
      >
        <div
          v-if="selectedRowLength"
          class="text-sm text-muted-foreground"
        >
          {{
            $t('components.pagination.selected', {
              selected: selectedRowLength,
              items: data?.length || 0,
            })
          }}
        </div>

        <ClientOnly>
          <UiPagination
            v-model:page="internalPage"
            :data-test-pagination="internalPage"
            :total="Math.max(1, (options?.pageCount || 0) * pagination.pageSize)"
            :items-per-page="pagination.pageSize"
            :disabled="loading"
            class="flex items-center gap-6 w-fit"
          >
            <div class="text-sm text-muted-foreground">
              {{
                $t('components.pagination.page', {
                  page: internalPage,
                  pages: Math.max(1, options?.pageCount || 0),
                })
              }}
            </div>
            <div class="flex items-center border-4 border-foreground bg-foreground/50 rounded-lg text-background">
              <UiPaginationFirst
                :disabled="!table.getCanPreviousPage()"
                class="border-0 border-r rounded-r-none border-r-foreground"
              />
              <UiPaginationPrev
                data-test-pagination-prev
                :disabled="!table.getCanPreviousPage()"
                class="border-0 border-r rounded-r-none border-r-foreground"
              />
              <UiPaginationNext
                data-test-pagination-next
                :disabled="!table.getCanNextPage() || (options?.pageCount && options.pageCount <= 1)"
                class="border-0 border-r rounded-r-none border-r-foreground"
              />
              <UiPaginationLast
                :disabled="!table.getCanNextPage() || (options?.pageCount && options.pageCount <= 1)"
                class="border-0"
              />
            </div>
          </UiPagination>

          <template #fallback>
            <div class="flex items-center gap-6 w-fit">
              <UiSkeleton class="w-16 h-5 rounded-full" />
              <UiSkeleton class="w-[136px] h-10" />
            </div>
          </template>
        </ClientOnly>
      </div>
    </div>

    <AnimationReveal>
      <button
        v-if="selectedRowLength"
        data-test-remove
        class="btn-destructive mt-2"
        :aria-label="$t('actions.bulkRemove', { number: selectedRowLength }, selectedRowLength)"
        @click="$emit('remove', table.getSelectedRowModel().rows.map(row => row.original.id))"
      >
        {{ $t('actions.bulkRemove', { number: selectedRowLength }, selectedRowLength) }}
      </button>
    </AnimationReveal>
  </div>
</template>
