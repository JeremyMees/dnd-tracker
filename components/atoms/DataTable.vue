<script setup lang="ts">
import { vAutoAnimate } from '@formkit/auto-animate'

defineEmits<{
  paginate: [value: number]
  remove: [items: number[]]
}>()

const props = withDefaults(
  defineProps<{
    items: any[]
    headers: TableHeader[]
    pages: number
    perPage: number
    totalItems: number
    type: 'campaign' | 'encounter' | 'homebrew' | 'note'
    loading?: boolean
    select?: boolean
    hasRights?: boolean
  }>(), {
    loading: false,
    select: false,
    hasRights: undefined,
  },
)

const sortBy = defineModel<string>('sortBy', { required: true })
const sortACS = defineModel<boolean>('acs', { default: true, required: true })
const page = defineModel<number>('page', { default: 0 })
const search = defineModel<string>('search', { default: '' })

const user = useAuthenticatedUser()

const selectedAll = ref<boolean>(false)
const selected = ref<any[]>([])
const detailRow = ref<number>()

defineExpose({
  toggleRow,
  toggleAll,
  toggleSort,
  toggleDetailRow,
  selected,
  detailRow,
})

function toggleSort(key: string): void {
  if (!props.items.length && props.loading) return

  sortACS.value = key === sortBy.value ? !sortACS.value : false
  sortBy.value = key
  selectedAll.value = false
  selected.value = []
}

function toggleDetailRow(id: number): void {
  if (detailRow.value === id) detailRow.value = undefined
  else detailRow.value = id
}

function toggleRow(row: any): void {
  selectedAll.value = false
  selected.value = toggleArray(row, selected.value)

  if (selected.value.length === props.items.length) selectedAll.value = true
}

function toggleAll(): void {
  const itemsWithRights = props.items.filter((item) => {
    if (props.hasRights !== undefined) return props.hasRights

    const owner = isOwner(item, user.value.id, true)

    return props.type === 'campaign'
      ? owner
      : owner || isAdmin(item, user.value.id, true)
  })

  if (selected.value.length === itemsWithRights.length) selected.value = []
  else selected.value = itemsWithRights
}
</script>

<template>
  <section class="inline-block overflow-x-auto overflow-y-hidden w-full">
    <div class="pb-2 flex justify-between gap-8">
      <FormKit
        v-model="search"
        type="search"
        suffix-icon="search"
        outer-class="$reset !pb-0 max-w-[300px]"
        inner-class="$remove:border-background border-secondary $remove:bg-muted bg-secondary/50"
      />
      <div class="flex gap-4 items-center">
        <slot name="header" />
      </div>
    </div>
    <div class="bg-secondary/50 border-4 border-secondary rounded-lg relative overflow-y-hidden">
      <table class="min-w-full">
        <thead>
          <tr class="border-b border-secondary">
            <th
              v-if="select"
              class="td"
            >
              <FormKit
                v-model="selectedAll"
                type="checkbox"
                :disabled="loading || !items.length"
                outer-class="$reset !pb-0"
                wrapper-class="$remove:mb-0"
                decorator-class="$remove:mr-2"
                @click="toggleAll"
              />
            </th>
            <th
              v-for="{ label, sort, id } in headers"
              :key="label"
              class="py-2"
              :class="{
                'cursor-pointer': sort,
                '!cursor-progress': loading,
                '!cursor-not-allowed': items.length === 0,
              }"
              @click="sort && toggleSort(id)"
            >
              <div
                class="flex gap-2 items-center w-fit px-2 rounded-lg transition-colors duration-300"
                :class="{ 'bg-secondary': id === sortBy }"
              >
                <span class="truncate text-muted-foreground hover:text-foreground transition-colors duration-300">
                  {{ label }}
                </span>
                <Icon
                  v-if="sort"
                  name="tabler:arrows-sort"
                  class="size-5 min-w-5 text-tertiary/50 transition-all duration-300"
                  :class="{
                    '!text-tertiary': sortBy === id,
                    'rotate-180': sortBy === id && !sortACS,
                  }"
                  aria-hidden="true"
                />
              </div>
            </th>
          </tr>
        </thead>
        <tbody v-auto-animate>
          <slot
            v-if="loading"
            name="loading"
          />
          <slot
            v-else
            :rows="items"
          />

          <template v-if="(!items || !items.length) && !loading">
            <tr class="py-20">
              <td
                :colspan="select ? headers.length + 1 : headers.length"
                class="py-20 px-5 font-bold"
              >
                <div class="max-w-prose mx-auto text-center text-muted-foreground">
                  <slot name="empty" />
                </div>
              </td>
            </tr>
          </template>
        </tbody>
      </table>
      <AnimationReveal>
        <div
          v-if="pages > 1"
          class="flex justify-end p-2"
        >
          <Pagination
            v-model:page="page"
            :pages="pages"
            :per-page="perPage"
            :loading="loading"
            @paginate="e => {
              selectedAll = false
              selected = []
              $emit('paginate', e)
            }"
          />
        </div>
      </AnimationReveal>
      <div class="inset-0 z-[-1] fancy-shadow" />
    </div>
    <AnimationReveal>
      <button
        v-if="selected && selected.length"
        class="btn-destructive mt-4"
        :aria-label="$t('actions.bulkRemove', { number: selected.length })"
        @click="() => {
          $emit('remove', selected.map(item => item.id))
          selectedAll = false
          selected = []
        }"
      >
        {{
          $t('actions.bulkRemove', {
            number: selected.length,
            type: $t(`general.${type}`, (selected || []).length).toLowerCase(),
          })
        }}
      </button>
    </AnimationReveal>
  </section>
</template>
