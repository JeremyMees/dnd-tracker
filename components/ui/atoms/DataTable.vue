<script setup lang="ts">
import { vAutoAnimate } from '@formkit/auto-animate'

defineEmits<{
  paginate: [value: number]
  remove: [items: number[]]
}>()

const props = withDefaults(
  defineProps<{
    items: any[]
    headers: TableHeader[]
    pages: number
    perPage: number
    totalItems: number
    type: 'campaigns' | 'encounters'
    loading?: boolean
    select?: boolean
  }>(), {
    loading: false,
    select: false,
  },
)

const sortedBy = defineModel<string>('sortedBy', { required: true })
const sortACS = defineModel<boolean>('acs', { default: false, required: true })
const page = defineModel<number>('page', { default: 0 })
const search = defineModel<string>('search', { default: '' })

const { t } = useI18n()
const profile = useProfile()

const selectedAll = ref<boolean>(false)
const selected = ref<any[]>([])

defineExpose({ toggleRow, toggleAll, toggleSort, selected })

function toggleSort(key: string): void {
  if (!props.items.length && props.loading) return

  sortACS.value = key === sortedBy.value ? !sortACS.value : false
  sortedBy.value = key
  selectedAll.value = false
  selected.value = []
}

function toggleRow(row: any): void {
  selectedAll.value = false
  selected.value = toggleArray(row, selected.value)

  if (selected.value.length === props.items.length) selectedAll.value = true
}

function toggleAll(): void {
  const itemsWithRights = props.items.filter((item) => {
    const owner = isOwner(item, profile.user!.id)

    return props.type === 'campaigns'
      ? owner
      : owner || isAdmin(item, profile.user!.id)
  })

  if (selected.value.length === itemsWithRights.length) selected.value = []
  else selected.value = itemsWithRights
}
</script>

<template>
  <section class="inline-block overflow-x-auto overflow-y-hidden w-full">
    <div class="bg-slate-700/50 border-4 border-slate-700 rounded-lg relative overflow-y-hidden">
      <div class="p-2 flex justify-between gap-8">
        <FormKit
          v-model="search"
          type="search"
          suffix-icon="search"
          outer-class="$reset !pb-0 max-w-[300px]"
        />
        <div class="flex gap-4 items-center">
          <slot name="header" />
        </div>
      </div>
      <table class="min-w-full">
        <thead>
          <tr class="border-b border-slate-700">
            <th
              v-if="select"
              class="td"
            >
              <FormKit
                v-model="selectedAll"
                type="checkbox"
                :disabled="loading || !items.length"
                outer-class="$reset !pb-0"
                wrapper-class="$remove:mb-0"
                decorator-class="$remove:mr-2"
                @click="toggleAll"
              />
            </th>
            <th
              v-for="{ label, sort, id } in headers"
              :key="label"
              class="py-2"
              :class="{
                'cursor-pointer': sort,
                '!cursor-progress': loading,
                '!cursor-not-allowed': items.length === 0,
              }"
              @click="sort && toggleSort(id)"
            >
              <div
                class="flex gap-2 items-center w-fit px-2 rounded-lg transition-colors duration-300"
                :class="{ 'bg-slate-700': id === sortedBy }"
              >
                <span class="text-slate-300 hover:text-white transition-colors duration-300">
                  {{ label }}
                </span>
                <Icon
                  v-if="sort"
                  name="ph:arrows-down-up-bold"
                  class="w-5 h-5 text-secondary/50 transition-all duration-300"
                  :class="{
                    '!text-secondary': sortedBy === id,
                    'rotate-180': sortedBy === id && !sortACS,
                  }"
                  aria-hidden="true"
                />
              </div>
            </th>
          </tr>
        </thead>
        <tbody v-auto-animate>
          <slot :rows="items" />

          <tr
            v-if="!items || !items.length"
            class="py-20"
          >
            <td
              :colspan="select ? headers.length + 1 : headers.length"
              class="py-20 px-5 font-bold"
            >
              <div class="max-w-prose mx-auto text-center">
                <slot name="empty" />
              </div>
            </td>
          </tr>
        </tbody>
      </table>
      <AnimationReveal>
        <div
          v-if="pages > 1"
          class="flex justify-end p-2"
        >
          <TablePagination
            v-model:page="page"
            :total-items="totalItems"
            :total-pages="pages"
            :loading="loading"
            :per-page="perPage"
            @paginate="e => {
              selectedAll = false
              selected = []
              $emit('paginate', e)
            }"
          />
        </div>
      </AnimationReveal>
      <div class="inset-0 z-[-1] fancy-shadow" />
    </div>
    <AnimationReveal>
      <button
        v-if="selected && selected.length"
        class="btn-danger mt-4"
        :aria-label="t('actions.bulkRemove', { number: selected.length })"
        @click="() => {
          $emit('remove', selected.map(item => item.id))
          selectedAll = false
          selected = []
        }"
      >
        {{ t('actions.bulkRemove', { number: selected.length, type: t(`general.${type}`).toLowerCase() }) }}
      </button>
    </AnimationReveal>
  </section>
</template>
