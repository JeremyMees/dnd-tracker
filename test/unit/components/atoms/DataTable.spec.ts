import { mountSuspended } from '@nuxt/test-utils/runtime'
import type { ColumnDef, Row, TableOptions } from '@tanstack/vue-table'
import { createColumnHelper } from '@tanstack/vue-table'
import { describe, expect, it, vi } from 'vitest'
import DataTable from '~/components/atoms/DataTable.vue'
import { expandButton, selectButton } from '~~/tables/generate-functions'

interface TestData {
  id: number
  name: string
  age: number
}

const mockData: TestData[] = [
  { id: 1, name: 'John Doe', age: 30 },
  { id: 2, name: 'Jane Smith', age: 25 },
  { id: 3, name: 'Bob Johnson', age: 40 },
]

const columnHelper = createColumnHelper<TestData>()

const mockColumns: ColumnDef<TestData>[] = [
  columnHelper.display({
    id: 'checkbox',
    header: '',
    cell: ({ row }) => row.getCanSelect()
      ? selectButton({ checked: row.getIsSelected(), cb: row.getToggleSelectedHandler(), disabled: false })
      : null,
  }),
  columnHelper.display({
    id: 'expand',
    header: '',
    cell: ({ row }) => expandButton({
      content: `actions.${row.getIsExpanded() ? 'hide' : 'show'}`,
      expanded: row.getIsExpanded(),
      cb: () => row.toggleExpanded(),
    }),
  }),
  columnHelper.accessor('id', {
    header: 'ID',
    cell: ({ row }) => row.getValue('id'),
  }) as ColumnDef<TestData>,
  columnHelper.accessor('name', {
    header: 'Name',
    cell: ({ row }) => row.getValue('name'),
  }) as ColumnDef<TestData>,
  columnHelper.accessor('age', {
    header: 'Age',
    cell: ({ row }) => row.getValue('age'),
  }) as ColumnDef<TestData>,
]

interface Props {
  columns: ColumnDef<any, any>[]
  data: any[]
  loading: boolean
  options?: Partial<TableOptions<any>>
  emptyMessage?: string
  permission?: boolean | ((item: any) => Promise<boolean>)
  expandedMarkup?: (row: Row<any>) => VNode
}

const props: Props = {
  columns: mockColumns,
  data: mockData,
  loading: false,
  options: { pageCount: 1 },
  emptyMessage: 'No data available',
  permission: true,
  expandedMarkup: row => h('div', { class: 'expanded-row' }, `Details for ${row.original.name}`),
}

describe('DataTable', () => {
  it('Should match snapshot', async () => {
    const component = await mountSuspended(DataTable, { props })
    expect(component.html()).toMatchSnapshot()
  })

  it('Should render table with correct data', async () => {
    const component = await mountSuspended(DataTable, { props })

    const headers = component.findAll('th')
    expect(headers.length).toBe(5)
    expect(headers[0]!.text()).toBe('')
    expect(headers[1]!.text()).toBe('')
    expect(headers[2]!.text()).toContain('ID')
    expect(headers[3]!.text()).toContain('Name')
    expect(headers[4]!.text()).toContain('Age')

    const rows = component.findAll('tbody tr')
    expect(rows.length).toBe(3)

    const firstRowCells = rows[0]!.findAll('td')
    expect(firstRowCells[2]!.text()).toContain('1')
    expect(firstRowCells[3]!.text()).toContain('John Doe')
    expect(firstRowCells[4]!.text()).toContain('30')

    expect(component.find('[data-test-empty]').exists()).toBeFalsy()
  })

  it('Should show loading state correctly', async () => {
    const component = await mountSuspended(DataTable, {
      props: { ...props, loading: true, data: [] },
      slots: {
        loading: () => h('div', 'Loading'),
      },
    })

    expect(component.html()).toContain('<div>Loading</div>')
  })

  it('Should show empty message when no data', async () => {
    const component = await mountSuspended(DataTable, { props: { ...props, data: [] } })

    const empty = component.find('[data-test-empty]')
    expect(empty.exists()).toBeTruthy()
    expect(empty.text()).toBe(props.emptyMessage)
  })

  it('Should emit remove event when bulk remove button is clicked', async () => {
    const component = await mountSuspended(DataTable, { props })

    const checkbox = component.find('button[role="checkbox"]')
    expect(checkbox.exists()).toBeTruthy()
    await checkbox.trigger('click')
    await nextTick()

    const removeButton = component.find('[data-test-remove]')
    expect(removeButton.exists()).toBeTruthy()
    await removeButton.trigger('click')

    const emitted = component.emitted()
    expect(emitted.remove).toBeTruthy()
    expect(emitted.remove![0]).toEqual([[1]])
  })

  it('Should handle row expansion correctly', async () => {
    const component = await mountSuspended(DataTable, { props })

    let expansionButton = component.find('button[arialabel="actions.show"]')
    expect(expansionButton.exists()).toBeTruthy()
    await expansionButton.trigger('click')
    await nextTick()

    let expandedContent = component.find('.expanded-row')
    expect(expandedContent.exists()).toBeTruthy()
    expect(expandedContent.text()).toContain('Details for John Doe')

    expansionButton = component.find('button[arialabel="actions.hide"]')
    expect(expansionButton.exists()).toBeTruthy()
    await expansionButton.trigger('click')
    await nextTick()

    expandedContent = component.find('.expanded-row')
    expect(expandedContent.exists()).toBeFalsy()
  })

  it('Should handle pagination correctly', async () => {
    const paginationProps = {
      ...props,
      options: {
        pageCount: 3,
        initialState: {
          pagination: {
            pageIndex: 0,
            pageSize: 1,
          },
        },
      },
    }

    const component = await mountSuspended(DataTable, { props: paginationProps })

    let paginationText = component.find('[data-test-pagination="1"]')
    expect(paginationText.exists()).toBeTruthy()

    const nextButton = component.find('button[data-test-pagination-next]')
    expect(nextButton.exists()).toBeTruthy()
    await nextButton.trigger('click')
    await nextTick()

    paginationText = component.find('[data-test-pagination="2"]')
    expect(paginationText.exists()).toBeTruthy()

    const prevButton = component.find('button[data-test-pagination-prev]')
    expect(prevButton.exists()).toBeTruthy()
    await prevButton.trigger('click')
    await nextTick()

    paginationText = component.find('[data-test-pagination="1"]')
    expect(paginationText.exists()).toBeTruthy()
  })

  it('Should handle custom permissions correctly', async () => {
    const permissionFn = vi.fn().mockImplementation(item => Promise.resolve(item.id === 1))

    const component = await mountSuspended(DataTable, {
      props: {
        ...props,
        permission: permissionFn,
      },
    })

    expect(permissionFn).toHaveBeenCalledTimes(3)
    expect(permissionFn).toHaveBeenCalledWith(mockData[0])
    expect(permissionFn).toHaveBeenCalledWith(mockData[1])
    expect(permissionFn).toHaveBeenCalledWith(mockData[2])

    const checkboxes = component.findAll('button[role="checkbox"]')
    expect(checkboxes.length).toBe(1)
  })
})
