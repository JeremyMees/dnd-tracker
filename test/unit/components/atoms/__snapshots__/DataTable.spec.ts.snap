// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataTable > Should match snapshot 1`] = `
"<div class="flex flex-col gap-2">
  <div class="flex flex-col-reverse sm:flex-row gap-2 justify-between items-center">
    <div id="formkit-stub" type="search" name="search" prefix-icon="tabler:search" outer-class="$remove:mb-4 w-full sm:w-auto" inner-class="$remove:border-background $remove:bg-muted border-secondary bg-secondary/50" value=""></div>
    <div class="flex justify-end w-full sm:w-auto"></div>
  </div>
  <div class="rounded-lg border-4 border-secondary bg-secondary/50">
    <div class="relative w-full overflow-auto">
      <table class="w-full caption-bottom text-sm">
        <thead class="[&amp;_tr]:border-b">
          <tr class="border-b border-secondary transition-colors data-[state=selected]:bg-secondary">
            <th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">
              <div class="flex items-center gap-2 w-fit">
                <!--v-if-->
              </div>
            </th>
            <th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">
              <div class="flex items-center gap-2 w-fit">
                <!--v-if-->
              </div>
            </th>
            <th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 right-0 cursor-pointer select-none" data-pinned="false">
              <div class="flex items-center gap-2 w-fit">ID
                <!--v-if-->
              </div>
            </th>
            <th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 right-0 cursor-pointer select-none" data-pinned="false">
              <div class="flex items-center gap-2 w-fit">Name
                <!--v-if-->
              </div>
            </th>
            <th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 right-0 cursor-pointer select-none" data-pinned="false">
              <div class="flex items-center gap-2 w-fit">Age
                <!--v-if-->
              </div>
            </th>
          </tr>
        </thead>
        <tbody class="[&amp;_tr:last-child]:border-0">
          <tr class="border-b border-secondary transition-colors data-[state=selected]:bg-secondary" data-state="false">
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false"><button class="peer size-5 shrink-0 rounded-lg border-2 border-foreground bg-foreground/50 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary/50 data-[state=checked]:border-primary data-[state=checked]:text-primary-foreground" role="checkbox" type="button" aria-checked="false" aria-required="false" data-state="unchecked">
                <!---->
                <!---->
              </button></td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false"><span data-v-tippy=""><button arialabel="actions.show" class="icon-btn-help"><span class="iconify i-tabler:chevron-right size-5 transition-transform duration-200 ease-in-out" aria-hidden="true" ariahidden="true"></span></button></span></td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">1</td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">John Doe</td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">30</td>
          </tr>
          <!--v-if-->
          <tr class="border-b border-secondary transition-colors data-[state=selected]:bg-secondary" data-state="false">
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false"><button class="peer size-5 shrink-0 rounded-lg border-2 border-foreground bg-foreground/50 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary/50 data-[state=checked]:border-primary data-[state=checked]:text-primary-foreground" role="checkbox" type="button" aria-checked="false" aria-required="false" data-state="unchecked">
                <!---->
                <!---->
              </button></td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false"><span data-v-tippy=""><button arialabel="actions.show" class="icon-btn-help"><span class="iconify i-tabler:chevron-right size-5 transition-transform duration-200 ease-in-out" aria-hidden="true" ariahidden="true"></span></button></span></td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">2</td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">Jane Smith</td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">25</td>
          </tr>
          <!--v-if-->
          <tr class="border-b border-secondary transition-colors data-[state=selected]:bg-secondary" data-state="false">
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false"><button class="peer size-5 shrink-0 rounded-lg border-2 border-foreground bg-foreground/50 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary/50 data-[state=checked]:border-primary data-[state=checked]:text-primary-foreground" role="checkbox" type="button" aria-checked="false" aria-required="false" data-state="unchecked">
                <!---->
                <!---->
              </button></td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false"><span data-v-tippy=""><button arialabel="actions.show" class="icon-btn-help"><span class="iconify i-tabler:chevron-right size-5 transition-transform duration-200 ease-in-out" aria-hidden="true" ariahidden="true"></span></button></span></td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">3</td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">Bob Johnson</td>
            <td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 right-0" data-pinned="false">40</td>
          </tr>
          <!--v-if-->
        </tbody>
      </table>
    </div>
    <div class="justify-end p-4 border-t border-secondary flex items-center gap-2">
      <!--v-if-->
      <nav data-test-pagination="1" class="flex items-center gap-6 w-fit">
        <div class="text-sm text-muted-foreground">components.pagination.page</div>
        <div class="flex items-center border-4 border-foreground bg-foreground/50 rounded-lg text-background"><button class="size-8 border-secondary rounded-lg p-0 flex items-center justify-center disabled:opacity-50 disabled:cursor-not-allowed border-0 border-r rounded-r-none border-r-foreground" aria-label="First Page" type="button" disabled=""><span class="iconify i-tabler:chevrons-left size-4 min-w-4" aria-hidden="true"></span></button><button class="size-8 border-secondary rounded-lg p-0 flex items-center justify-center disabled:opacity-50 disabled:cursor-not-allowed border-0 border-r rounded-r-none border-r-foreground" aria-label="Previous Page" type="button" disabled="" data-test-pagination-prev=""><span class="iconify i-tabler:chevron-left size-4 min-w-4" aria-hidden="true"></span></button><button class="size-8 border-secondary rounded-lg p-0 flex items-center justify-center disabled:opacity-50 disabled:cursor-not-allowed border-0 border-r rounded-r-none border-r-foreground" aria-label="Next Page" type="button" disabled="" data-test-pagination-next=""><span class="iconify i-tabler:chevron-right size-4 min-w-4" aria-hidden="true"></span></button><button class="size-8 border-secondary rounded-lg p-0 flex items-center justify-center disabled:opacity-50 disabled:cursor-not-allowed border-0" aria-label="Last Page" type="button" disabled=""><span class="iconify i-tabler:chevrons-right size-4 min-w-4" aria-hidden="true"></span></button></div>
      </nav>
    </div>
  </div>
  <div class="overflow-hidden">
    <!---->
  </div>
</div>"
`;
